services/src/main/java/com/linkedin/kmf/services/TopicRebalancer.java

```diff
@@ -0,0 +1,427 @@
+ * file except in compliance with the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package com.linkedin.kmf.services;
+
+import com.linkedin.kmf.services.configs.TopicRebalancerConfig;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import kafka.admin.AdminUtils;
+import kafka.admin.PreferredReplicaLeaderElectionCommand;
+import kafka.cluster.Broker;
+import kafka.common.TopicAndPartition;
+import kafka.utils.ZkUtils;
+import org.apache.kafka.common.Node;
+import org.apache.kafka.common.PartitionInfo;
+import org.apache.kafka.common.errors.InterruptException;
+import org.apache.kafka.common.security.JaasUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static com.linkedin.kmf.common.Utils.ZK_CONNECTION_TIMEOUT_MS;
+import static com.linkedin.kmf.common.Utils.ZK_SESSION_TIMEOUT_MS;
+
+import scala.collection.Seq;
+
+
+/**
+ * Runs this periodically to rebalance the monitored topic across brokers and to reassign _electedLeaders to brokers so that the
+ * monitored topic is sampling all the brokers evenly.
+ */
+public class TopicRebalancer implements Service  {
+
+  private static final Logger LOG = LoggerFactory.getLogger(TopicRebalancer.class);
+
+  /**
+   * The state of the monitored topic.
+   */
+  static class TopicState {
+
+    Map<Integer, Integer> _brokerToPartitionCount = new HashMap<>();
+    //Assigned _electedLeaders tracks if there is some partition for which it is the preferred leader which could be none.
+    Set<Integer> _preferredLeaders = new HashSet<>();
+    Set<Integer> _electedLeaders = new HashSet<>();
+    Collection<Broker> _allBrokers;
+    List<PartitionInfo> _partitionInfo;
+
+    TopicState(Map<Integer, Integer> brokerToPartitionCount, Set<Integer> electedLeaders,
+      Set<Integer> preferredLeaders, List<PartitionInfo> partitionInfo, Collection<Broker> allBrokers) {
+      this._brokerToPartitionCount = brokerToPartitionCount;
+      this._preferredLeaders = preferredLeaders;
+      this._electedLeaders = electedLeaders;
+      this._allBrokers = allBrokers;
+      this._partitionInfo = partitionInfo;
+    }
+
+    /**
+     *
+     * @param expectedRatio the expected partition to broker ratio
+     * @return true if the current partition to broker ratio is less than the expected ratio
+     */
+    boolean partitionsLow(double expectedRatio) {
```

Comment:
I think it is better for `partitionsLow`, `brokerMissingPartition` and `brokerNotElectedLeader` to have the same style. Eitehr they are all stateful (i.e. read state from class private variables ) or stateless (i.e. read state from method input). Maybe we can change the method signature to be `partitionLow()`